### PR TITLE
kv/bulk: skip exact matches in SSTBatcher when ingestAll is true

### LINKED ISSUE #83740: ccl/streamingccl/streamingest: TestRandomClientGeneration failed
ccl/streamingccl/streamingest.TestRandomClientGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5628464?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5628464?buildTab=artifacts#/) on master @ [4e8134033890c956d39020f28a0e99b21b32649f](https://github.com/cockroachdb/cockroach/commits/4e8134033890c956d39020f28a0e99b21b32649f):

```
=== RUN   TestRandomClientGeneration
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestRandomClientGeneration1869708782
    test_log_scope.go:80: use -show-logs to present logs inline
    stream_ingestion_processor_test.go:464: unexpected meta error flushing: addsstable [/Tenant/30/Table/3/1/52/2/1,/Tenant/30/Table/53/1/96/0/NULL): SST stats are incorrect: diff(given, actual) = [KeyBytes: 5924 != 5036 ValBytes: 3523 != 2989 ValCount: 395 != 321]
    panic.go:661: -- test log scope end --
--- FAIL: TestRandomClientGeneration (30.77s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/bulk-io
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomClientGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### LINKED ISSUE #83741: ccl/streamingccl/streamingest: TestStreamIngestionFrontierProcessor failed
ccl/streamingccl/streamingest.TestStreamIngestionFrontierProcessor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5628464?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5628464?buildTab=artifacts#/) on master @ [4e8134033890c956d39020f28a0e99b21b32649f](https://github.com/cockroachdb/cockroach/commits/4e8134033890c956d39020f28a0e99b21b32649f):

```
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_proposal.go:766
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evalAndPropose
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:113
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeWriteBatch
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_write.go:168
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:481
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithoutRangeID
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:181
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:103
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_send.go:197
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).Send
          | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:191
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
          | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1007
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
          | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1056
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:723
          | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1
          | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:95
          | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:794
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:264
          | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:794
          | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:73
          | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:794
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:233
          | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
          | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:341
          | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:231
          | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:794
          | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func2
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:733
          | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ClientInterceptor.func2
          | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:233
          | github.com/cockroachdb/cockroach/pkg/rpc.getChainUnaryInvoker.func1
          | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:878
        Wraps: (7) SST stats are incorrect: diff(given, actual) = [GCBytesAge: 39761756016 != 0 KeyBytes: 32 != 20 ValBytes: 24 != 12 ValCount: 2 != 1]
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *assert.withAssertionFailure (6) *errbase.opaqueWrapper (7) *errutil.leafError TraceData:[] LeafTxnFinalState:<nil> RowNum:<nil> SamplerProgress:<nil> BulkProcessorProgress:<nil> Metrics:<nil>}
    --- FAIL: TestStreamIngestionFrontierProcessor/no-checkpoints (0.10s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/bulk-io
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionFrontierProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-17231

### LINKED ISSUE #83742: ccl/streamingccl/streamingest: TestStreamIngestionProcessor failed
ccl/streamingccl/streamingest.TestStreamIngestionProcessor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5628464?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5628464?buildTab=artifacts#/) on master @ [4e8134033890c956d39020f28a0e99b21b32649f](https://github.com/cockroachdb/cockroach/commits/4e8134033890c956d39020f28a0e99b21b32649f):

```
=== RUN   TestStreamIngestionProcessor
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestStreamIngestionProcessor1436480695
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestStreamIngestionProcessor
    stream_ingestion_processor_test.go:302: -- test log scope end --
--- FAIL: TestStreamIngestionProcessor (35.72s)
=== RUN   TestStreamIngestionProcessor/finite_stream_client
    row_buffer.go:216: unexpected metadata: &{[] flushing: addsstable [/Tenant/30"key_1",/Tenant/30"key_1\x00"): SST stats are incorrect: diff(given, actual) = [GCBytesAge: 119285270856 != 0 KeyBytes: 56 != 20 ValBytes: 48 != 12 ValCount: 4 != 1] [] <nil> <nil> <nil> <nil> <nil>}
    --- FAIL: TestStreamIngestionProcessor/finite_stream_client (0.09s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/bulk-io
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamIngestionProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by stevendanna**
This changes the semantics of ingestAll to skip ingesting exact
matches. We expect to see such exact matches when processing duplicate
events from a tenant replication stream.

Previously, such exact matches didn't cause a problem other than
wasted work.

However, recently a number of streaming tests are recently failing
under race with

    stream_ingestion_processor_test.go:464: unexpected meta error
    flushing: addsstable
    [/Tenant/30/Table/3/1/52/2/1,/Tenant/30/Table/53/1/96/0/NULL): SST
    stats are incorrect: diff(given, actual) = [KeyBytes: 5924 != 5036
    ValBytes: 3523 != 2989 ValCount: 395 != 321]

When under race, AddSST re-calculates any MVCCStats sent to
it. However, it does so using a different iterator than we do in
SSTBatcher. The new iterators used during the check treats duplicates
in the SST differently -- leading to a difference in the stats
computation.

The difference in iterator behaviour is arguable a bug (https://github.com/cockroachdb/cockroach/issues/83766); but
since the duplicates serve no purpose, removing them during batching
seems prudent.

Fixes https://github.com/cockroachdb/cockroach/issues/83740
Fixes https://github.com/cockroachdb/cockroach/issues/83741
Fixes https://github.com/cockroachdb/cockroach/issues/83742

Release note: None

### CODE REVIEW

**Comment by erikgrinaker** [comment]
This just seems wrong. Unless the SST actually contains any MVCC range tombstones, this should bot affect stats. I'll have a look in the morning, sorry for the trouble!

**Comment by erikgrinaker** [comment]
Why do the SSTs contain duplicates though? Is that ever desired or useful?

I think this might simply be because the new SST iterators just use a plain ol' `pebbleIterator`, which assumes there can't be duplicates (Pebble never emits duplicates in regular iterators).

**Comment by stevendanna** [comment]
> Why do the SSTs contain duplicates though? Is that ever desired or useful?

Good question.

Over in stream ingestion, the tests simulate duplicates being sent.  The sst batcher has a feature to skip duplicates that it sees, but for some reason we don't set that and instead set ingestAll which explicitly allows the duplicates.  But, it's not clear to me why we would want this -- it's just more data we don't need to be shovelling around.  I'll switch us over to using skipDuplicates rather than injestAll and see if anything fails.

**Comment by stevendanna** [comment]
Ahh, actually the original commit sheds some light on it: https://github.com/cockroachdb/cockroach/commit/baf89289c7f

**Comment by stevendanna** [comment]
I'll look at the sst batcher side a bit more tomorrow. I think something like what I've just pushed is what we may want, but I need to write some tests.

**Comment by stevendanna** [comment]
@erikgrinaker I moved the iterator issue to #83766 for further discussion and changed this PR to avoid these exact duplicates.

File: pkg/kv/bulk/sst_batcher.go [hunk 1]
```diff
@@ -137,13 +137,14 @@ type SSTBatcher struct {
...
-	batchEndKey     []byte
-	batchEndValue   []byte
-	flushKeyChecked bool
-	flushKey        roachpb.Key
+	sstWriter         storage.SSTWriter
+	sstFile           *storage.MemFile
+	batchStartKey     []byte
+	batchEndKey       []byte
+	batchEndValue     []byte
+	batchEndTimestamp hlc.Timestamp
```
Comment (erikgrinaker) [inline pkg/kv/bulk/sst_batcher.go:145]: nit: at this point, we could just as well use a `storage.MVCCKeyValue` to gather all of these.

**Review by erikgrinaker — COMMENTED** [review]
If I'm understanding this correctly, we'll now skip exact duplicates, but still include duplicate MVCC keys. I.e. we'll happily put `a@5=foo` and `a@5=bar` in the same SST. The server will simply discard `a@5=bar` in this case, so I'm not sure if this behavior is useful, just pointing it out.

**Comment by erikgrinaker** [comment]
It may be better to fail loudly, since I would say something is very wrong in this case. Otherwise, we risk data loss (we'll discard the second value, and it isn't clear whether the first or the second value is the correct one). But I don't know what the motivation was for this behavior in the first place.

**Comment by stevendanna** [comment]
I updated this to fail in the case of a key-timestamp match where the values differ. I also increased some of the test coverage and unksipped some tests that were skipped because of this.

**Review by erikgrinaker — APPROVED** [review]
Thanks!

**Comment by stevendanna** [comment]
bors r=erikgrinaker

**Comment by stevendanna** [comment]
One of the failures is in LogicTest:

```
------- Stdout: -------
=== RUN   TestLogic
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic3570340496
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestLogic
    logic.go:4532: -- test log scope end --
    logic_test.go:33: Leaked goroutine: goroutine 8490512 [select]:
        github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked.func1()
          /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/vfs/disk_health.go:302 +0xe9
        created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked
          /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/vfs/disk_health.go:297 +0x85
        Leaked goroutine: goroutine 8490786 [select]:
        github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked.func1()
          /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/vfs/disk_health.go:302 +0xe9
        created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked
          /go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/vfs/disk_health.go:297 +0x85
--- FAIL: TestLogic (2013.60s)
```

The other is in testccl/workoad/schecmachange.TestWorkload.  I'm running that test under stress for a while to see if I can see anything similar.

**Comment by stevendanna** [comment]
bors retry

**Comment by nicktrav** [comment]
The logictest failure look like it might be on us (Storage) - I think there may be a few of these goroutine leaks in tests coming from the FS code.

**Comment by jbowens** [comment]
The goroutine leak is likely from using an Engine's FS after the Engine has already been closed.

**Pull request merged**